### DISCUSSION TITLE
Raspberry Pi 4 Issues

### DISCUSSION THREAD

**Discussion by wellmans** [kenschneider18/rpi-metro-display#9] [General]
I tried. I really tried but I can't get anything to display. I got the demos from Adafruit to work fine. Since I'm not using dietpi I have some reservations about the paths. I'm confused what what form the run.sh is supposed to be and what to do with the <station-code> <direction-code> placeholders and the metro-display.service entries such as After=. The service starts or so it seems but there is nothing happening or any log file. I literally replaced the placeholders with:

#!/bin/bash
python /home/pi/metro-sign/rpi-rgb-led-matrix/rpi-metro-display/rpi-metro-display.py /home/pi/metro-sign/rpi-rgb-led-matrix/rpi-metro-display/log.txt /home/pi/metro-sign/rpi-rgb-led-matrix/rpi-metro-display/6x10.bdf /home/pi/metro-sign/rpi-rgb-led-matrix/rpi-metro-display/lines.json /home/pi/metro-sign/rpi-rgb-led-matrix/rpi-metro-display/stations.json.

Any thoughts would be appreciated. I know enough about this stuff to be dangerous but not enough to infer things or read between the lines.

Thanks,
Stephen

EDIT: I stepped through run.sh and ran rpi-metro-display.py manually and got the display to work. I used a station code E02 and the direction 0. I'm getting a display now but not sure what the problem is/was. It's after 11 so maybe the station data is available. I'm getting closer but still appreciate comments. My display is all sparkly especially with activity on the monitor but I'm using convenience. I have to push-fit headers coming to I can use jumpers instead of soldering.

EDIT2: It's working executing the run.sh manually. I still have sparkles and am going to  work with the slowdown and then try to get the whole project to start with the service. 

I failed to mention I'm using the bonnet w/o RTC. The bonnet still uses the hat setting? The bonnet has a performance solder I'm going to try to use w/o soldering. I am going to see if I need to rebuild the drivers. I'm using a standard install of RPi 32 bit fwiw. 

So, most of the primary post is moot but I'm not sure if I should edit the entire post or just amend the original like this.

I just wanted to say thanks and bravo for this very nice project.

**Comment by kenschneider18** [comment]
Glad you were able to get it running manually!

Let's start with having it run via `systemctl`. The `metro-display.service` has some dietpi specific stuff baked in, however, it can be modified to work with any OS running `systemd`. I'll assume that you're using some variant of Raspbian. In that case, you'll need to remove `dietpi-boot.service dietpi-postboot.service` from [line 3 of `metro-display.service`](https://github.com/kenschneider18/rpi-metro-display/blob/master/metro-display.service#L3). Then also fix [the `ExecStart` line](https://github.com/kenschneider18/rpi-metro-display/blob/master/metro-display.service#L9) to be equal to the canonical path of your `run.sh`. You can get the canonical path by running `pwd` from the directory where your `run.sh` is located. Then ensure that's copied to the `systemd` directory from step 12 (if you didn't edit it in place). That should hopefully do it for you.

On to the sparkles, I can't see exactly what you're talking about obviously but I'm guessing it's a performance issue. I'm once again assuming you're using some sort of Raspbian. By default, I believe Raspbian boots into a GUI desktop environment which puts some strain on the system thus limiting the performance of your display. The first thing I'd try is disabling that GUI and running through the [troubleshooting steps from `rpi-rgb-led-matrix`](https://github.com/hzeller/rpi-rgb-led-matrix#troubleshooting).

Assuming that doesn't work, you might need to change some configuration of the program. Check out [this section](https://github.com/hzeller/rpi-rgb-led-matrix#misc-options) of the `rpi-rgb-led-matrix` README. One part shows "ghosting" specifically, but many of these configs are values I've hard coded in my program. [Here's where they live in my code](https://github.com/kenschneider18/rpi-metro-display/blob/master/rpi-metro-display.py#L104-L109). I'd play around with these based on the linked section of hzeller's README and see if that helps.

You can try the performance jumper but definitely don't solder it permanently unless you're sure it works. You could also try [this](https://github.com/hzeller/rpi-rgb-led-matrix#improving-flicker) but when I tried it just made things worse. Not to mention that if you do this, you have to modify the library's `Makefile` and re-compile it (step 5 in my README).

Let me know if this helps!

**Comment by wellmans** [comment]
Thanks for the response. As for the sparkle, I edit the code and changed the slowdown to 4 and it looks fantastic. I was going to jump pins to get quality, which loses sound, but it looks so good I don't see why. I do have a question about why chain=4 in the code but it doesn't seem to be problematic. 

I did edit the service file and moved it over. I also edited in place at systemctl. I'm not sure if that's good or bad. I did place the path in the service file and played with it a bit but I still had issues getting it to start. It starts but doesn't launch run.py. I will step it back and go through the steps again to see what happens. I'm not actually interested in booting to the display as it a 4 with 4h and want my son, who has an apt in DC, have access to its full capability. 

I'm working on ssh to show him how to change the station and direction but he's usually only interested in one station. 

I plan to pick through your code to see if I can display a notice that it's after hours. Maybe even a different display like weather when there's no data. 

It looks great. I have to frame the panels and work on a power/cable solution to place the display away from the pi and put it in a case.

Thanks again for the response. I'm new to all this but I can pick through the code and learn from it. It's a win-win for old retired me.

**Reply by kenschneider18** [reply]
Awesome, glad it was a simple fix. That's to do with the library I use to drive the display. I believe that 1 unit of display is a 32x32 matrix. The 32x128 display then is 4x.

Any reason you don't want to use the API to switch station/direction?

There's definitely room to improve upon what's already there. I'm open to contributions if you're interested in having that become part of the main code.

Again, very glad to hear you have it working and are enjoying it!

**Reply by wellmans** [reply]
I am using the API (I think) with the curl commands you documented. 

I really need to fully understand the service. I can get it to start but it doesn't launch the run.sh. In fact, I don't have a full grasp on the scripting and running scrips command line. I can't run run.sh but I can copy the contents and use "sudo python" to get it running once the service has started. A lot of my issues is learning and every bit takes me down a rabbit hole.

Again, I'm not interested in the pi booting straight to the display but trying to get it to the point my son can use it - he has about zero experience in programming but is OK dropping operating systems on his devices. 

Thanks again.

**Reply by kenschneider18** [reply]
Ah, yeah I was looking at it from the angle of your son needing to know how to `ssh` which for just controlling the sign he won't, but if it's not auto-running that's still necessary.

**Reply by patrickhills** [reply]
I just wanted to chime in here with a +1 for changing slowdown to 4. I had 'sparkles' as well until finding this suggesting. Might be worth adding to the troubleshooting section?

Thanks for publishing this amazing product!

**Reply by wellmans** [reply]
Patrick, it is amazing! If/when I get up to speed programming on the pi I will want to add a GUI to change the station and/or direction on the fly.

**Comment by wellmans** [comment]
I'm embarrassed to say that it was run.sh didn't have execute permissions. Now it boots directly to the display. Oh well ... we live and learn ... hopefully.

**Reply by kenschneider18** [reply]
I should be the embarrassed one. Git retains file permissions and I definitely thought I'd committed `run.sh` with exec permissions. Thanks for finding that!

**Comment by ericsussman** [comment]
I'm a little late to the party (and very much a beginner), but am very excited about this project! However, I am also having difficulties running this on my Pi4 with Python2. I've walked through the install instructions on DietPi exactly how they are outlines, but haven't been able to produce any sort of display (my line.json and stations.json files are fully populated though). When I try to run `run.sh` manually, I still can't display anything. I've tested Adafruit's display demo with my system and that worked without any problems. 

I've outlined a few notes on the errors I've been receiving below:

- When installing the pre-requisite software in Step 2, I receive an error if I try to do so using the provided code block, so I've been using Python 3. The error message I receive is the following:
`Note, selecting 'python-is-python2' instead of 'python'. Package python-pip is not available, but is referred to by another package. This may mean that the package is missing, has been obsoleted, or is only availabel from another source. However the following packages replace it: python3-pip`

- After installation using Python3, I must manually install the packages `rgbmatrix` and `libopenjp2-7`
- When I try to run run.sh manually, I get the following error message:
`Traceback (most recent call last):
File "/home/dietpi/metro-sign/rpi-metro-display/rpi-metro-display.py", line 27, in <module> from incidents import get_incidents, draw_incident
File "/home/dietpi/metro-sign/rpi-metro-display/inscidents.py", line 42
logging.info("Attempting to get train data!")
TabError: inconsistent use of tabs and spaces in indentation`

Any advice? Thanks!

**Reply by kenschneider18** [reply]
That's a good note with the Python 3 stuff. Honestly at this point it's probably a good thing that you're using Python 3 and I should update the README with Python 3 instructions. Python 2 isn't supported anymore so everything should be moving that direction anyway.

The error you're seeing with `run.sh` doesn't look like it's related to Python version. Is there any chance you've modified the `incidents.py` file? Python is very sensitive to whether tabs or spaces are used in its files. If there's a mix of tabs and spaces, it will fail to compile with an error like the above.

**Reply by kenschneider18** [reply]
Okay, so I just re-cloned the repo myself to address another issue that's come up and I see exactly where the problem is. I'm honestly not sure how this code has ever compiled/worked. Line 42 in Incidents is definitely using a tab for indentation where it should be using spaces. I'll be pushing that update tonight and cutting a new release.

**Reply by ericsussman** [reply]
Thanks for looking into this more! I now receive no error messages when using the new code, but after stepping through the instructions again and trying to run the new code either through `sudo systemctl start metro-display` or manually running run.sh I get nothing appearing on my display. Any idea what could be wrong?

**Reply by wellmans** [reply]
> Thanks for looking into this more! I now receive no error messages when using the new code, but after stepping through the instructions again and trying to run the new code either through `sudo systemctl start metro-display` or manually running run.sh I get nothing appearing on my display. Any idea what could be wrong?

Just to omit the obvious are you sure the panel works? Have you tested it yet?

**Reply by ericsussman** [reply]
> > Thanks for looking into this more! I now receive no error messages when using the new code, but after stepping through the instructions again and trying to run the new code either through `sudo systemctl start metro-display` or manually running run.sh I get nothing appearing on my display. Any idea what could be wrong?
> 
> Just to omit the obvious are you sure the panel works? Have you tested it yet?

Yes, I've tested my display setup with Adafruit's display demo and that worked without any problems.

**Reply by kenschneider18** [reply]
Is there any output in your `log.txt` file? You mentioned you're using python3, right?

**Reply by ericsussman** [reply]
I am using python3, but the only change has been to use `pip3` and `python3` in place of `pip` and `python`. Below is my `log.txt` file.:

```
  GNU nano 5.4                         log.txt                                  
Type: <class 'TypeError'>
Value: one character bytes, bytearray or integer expected
TB: <traceback object at 0xb586adc8>
Uncaught exception!
Type: <class 'TypeError'>
Value: one character bytes, bytearray or integer expected
TB: <traceback object at 0xb57adf88>
Uncaught exception!
Type: <class 'TypeError'>
Value: one character bytes, bytearray or integer expected
TB: <traceback object at 0xb5800f88>
Uncaught exception!
Type: <class 'TypeError'>
Value: one character bytes, bytearray or integer expected
TB: <traceback object at 0xb57bede8>
Uncaught exception!
Type: <class 'TypeError'>
Value: one character bytes, bytearray or integer expected
TB: <traceback object at 0xb5801f48>
```

**Reply by kenschneider18** [reply]
Yeah, it's a Python 3 thing. I'm actually working on deprecating Python 2 as we speak. If you check out the `update/python3` branch you'll get a more descriptive error message something something `c_char`. I'm hoping to take a closer look tomorrow but seems like Python's Multiprocessing library changed the way it worked for Python 3.

I hoped this would work with Python 3 right out of the gate but alas, no.

**Reply by kenschneider18** [reply]
Okay so I actually got to it tonight. There's definitely a bug somewhere in displaying incidents I have to hunt down, but I've got a running display now. Try out the branch `update/python3` and let me know if you can start seeing train times. That should work now.

From the `rpi-metro-display` folder:
```sh
git checkout update/python3
```

To get back to the main branch:
```sh
git checkout master
```

**Reply by kenschneider18** [reply]
You might have to stash your changes to `run.sh` first:

```sh
git stash
```

Then checkout the branch, then run:

```sh
git stash pop
```

It will probably complain about a merge conflict. Just open the file with `nano` and remove all the lines other than your original run command and the `#!/bin/sh`.

I'm re-thinking how I distribute a run script template so this doesn't happen in the future.

**Reply by kenschneider18** [reply]
Incidents should be working now as well, I haven't tried out the REST API just yet but I'd be willing to bet there's going to be something in there I need to fix. At least now you should be able to run it and see all the data for the station and direction in `run.sh`

**Reply by ericsussman** [reply]
I've tried the new branch and am still getting a similar `log.txt` file with the same level of description (but with more errors logged). Do you think my problem has something to do with my `run.sh` file, copied below with api_key removed? 

`python3 /home/dietpi/metro-sign/rpi-metro-display/rpi-metro-display.py /home/dietpi/metro-sign/rpi-metro-display/log.txt <api_key>  E04 E10 /home/dietpi/metro-sign/rpi-metro-display/6x10.bdf /home/dietpi/metro-sign/rpi-metro-display/lines.json /home/dietpi/metro-sign/rpi-metro-display/stations.json`

**Reply by kenschneider18** [reply]
It may not be the only problem, but I do see one issue there. If I'm understanding correctly from what you're station and direction are, you're trying to load times at Columbia Heights in the direction of Greenbelt. You're totally good on the first part `E04`, however, the next variable should be the "group" number rather than the terminal station code.

In my experience the best way to figure out which group you're looking for is to just call this endpoint:

```sh
curl --location --request GET 'https://api.wmata.com/StationPrediction.svc/json/GetPrediction/E04' \
--header 'api_key: <api_key>'
```

And see which group corresponds to the direction you're looking for. In your case for Greenbelt, this is `1`. So your `run.sh` should read:

```sh
python3 /home/dietpi/metro-sign/rpi-metro-display/rpi-metro-display.py /home/dietpi/metro-sign/rpi-metro-display/log.txt <api_key> E04 1 /home/dietpi/metro-sign/rpi-metro-display/6x10.bdf /home/dietpi/metro-sign/rpi-metro-display/lines.json /home/dietpi/metro-sign/rpi-metro-display/stations.json
```

**Reply by ericsussman** [reply]
Yup, that was definitely an issue! I also redid the installation procedure and I got the display to work this time and it looks great! Thanks for all the help!

**Reply by ericsussman** [reply]
Is there a way to change the brightness of the rgb display? It looks like it has to be done when defining the color variables in `rpi-metro-display.py`, but I just wanted to double check.

**Reply by kenschneider18** [reply]
There is a setting in the [rpi-rgb-led-matrix](https://github.com/hzeller/rpi-rgb-led-matrix) library if I remember correctly. I tried playing with it and it ended up screwing with the colors and I never put much time into debugging it.
